// @aws-amplify/platform-core/lib/telemetry-data/telemetry_data.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.telemetryPayloadSchema = void 0;
const zod_1 = __importDefault(require("zod"));
const identifiersSchema = zod_1.default.object({
    payloadVersion: zod_1.default.string(),
    sessionUuid: zod_1.default.string(),
    eventId: zod_1.default.string(),
    timestamp: zod_1.default.string(),
    localProjectId: zod_1.default.string(),
    accountId: zod_1.default.string().optional(),
    awsRegion: zod_1.default.string().optional(),
});
const eventSchema = zod_1.default.object({
    state: zod_1.default.enum(['ABORTED', 'FAILED', 'SUCCEEDED']),
    command: zod_1.default.object({
        path: zod_1.default.array(zod_1.default.string()),
        parameters: zod_1.default.array(zod_1.default.string()),
    }),
});
const environmentSchema = zod_1.default.object({
    os: zod_1.default.object({
        platform: zod_1.default.string(),
        release: zod_1.default.string(),
    }),
    shell: zod_1.default.string(),
    npmUserAgent: zod_1.default.string(),
    ci: zod_1.default.boolean(),
    memory: zod_1.default.object({
        total: zod_1.default.number(),
        free: zod_1.default.number(),
    }),
});
const projectSchema = zod_1.default.object({
    dependencies: zod_1.default
        .array(zod_1.default.object({
        name: zod_1.default.string(),
        version: zod_1.default.string(),
    }))
        .optional(),
});
const latencySchema = zod_1.default.object({
    total: zod_1.default.number(),
    init: zod_1.default.number().optional(),
    synthesis: zod_1.default.number().optional(),
    deployment: zod_1.default.number().optional(),
    hotSwap: zod_1.default.number().optional(),
});
const errorSchema = zod_1.default.lazy(() => zod_1.default.object({
    name: zod_1.default.string(),
    message: zod_1.default.string(),
    stack: zod_1.default.string(),
    cause: zod_1.default.optional(errorSchema), // Recursive reference
}));
exports.telemetryPayloadSchema = zod_1.default.object({
    identifiers: identifiersSchema,
    event: eventSchema,
    environment: environmentSchema,
    project: projectSchema,
    latency: latencySchema,
    error: zod_1.default.optional(errorSchema),
});
